controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -210,7 +256,210 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
+            return streamMetadataStore.checkReaderGroupExists(scope, rgName)
+               .thenCompose(exists -> {
+               if (!exists) {
+                  return CompletableFuture.completedFuture(UpdateReaderGroupStatus.Status.RG_NOT_FOUND);
+               }
+               //2. check for generation && ID match with existing config
+               return streamMetadataStore.getReaderGroupConfigRecord(scope, rgName, context, executor)
+                      .thenCompose(rgConfigRecord -> {
+                        if (rgConfigRecord.getObject().getGeneration() != config.getGeneration()) {
+                            return CompletableFuture.completedFuture(UpdateReaderGroupStatus.Status.INVALID_CONFIG);
+                        }
+                        if (!rgConfigRecord.getObject().isUpdating()) {
+                          return streamMetadataStore.getReaderGroupId(scope, rgName, context, executor)
+                             .thenCompose(rgId -> {
+                             if (!config.getReaderGroupId().equals(rgId)) {
+                                 return CompletableFuture.completedFuture(UpdateReaderGroupStatus.Status.INVALID_CONFIG);
+                             }
+                             ImmutableSet<String> removeStreams = getStreamsToBeUnsubscribed(rgConfigRecord.getObject().getStartingStreamCuts().keySet(),
+                             config.getStartingStreamCuts().keySet().stream().map(s -> s.getScopedName()).collect(Collectors.toSet()));
+                             UpdateReaderGroupEvent event = new UpdateReaderGroupEvent(scope, rgName, requestId, rgId,
+                                                                 rgConfigRecord.getObject().getGeneration() + 1, removeStreams);
+                             //3. Create Reader Group Metadata and submit event
+                             return eventHelper.addIndexAndSubmitTask(event,
+                                    () -> streamMetadataStore.startRGConfigUpdate(scope, rgName, config, null, executor))
+                                          .thenCompose(x -> eventHelper.checkDone(() -> isRGUpdated(scope, rgName, executor))
+                                              .thenApply(done -> UpdateReaderGroupStatus.Status.SUCCESS));
+                             });
+                        } else {
+                          log.warn("Reader group update failed as another update is in progress.");
+                          return CompletableFuture.completedFuture(UpdateReaderGroupStatus.Status.FAILURE);
+                        }
+                      });
+               });
+        }, e -> Exceptions.unwrap(e) instanceof RetryableException, READER_GROUP_OPERATION_MAX_RETRIES, executor);
+    }
+
+    private ImmutableSet<String> getStreamsToBeUnsubscribed(final Set<String> currentConfigStreams, final Set<String> newConfigStreams) {
+        return ImmutableSet.copyOf(currentConfigStreams.stream()
```

Comment:
this is suboptimal as copyOf will make another pass.. better idea is to create an ImmutableSetBuilder.
then do foreach on each element and those that dont satisfy the filter predicate, add them to the builder.

Reply:
fixed